### DISCUSSION TITLE
Why EdgeX Store&Foreward does not guarantee messages are sent in order?

### DISCUSSION THREAD

**Discussion by amunozh** [edgexfoundry/community#271] [Q&A]
According to Store&Forward documentation, EdgeX does not guarantee messages are sent in the same order they were received. Why? How Store&Forward is implemented? Is there a way to guarantee the order?
![image](https://github.com/edgexfoundry/community/assets/38354345/959b9a77-77ae-42d5-bf9f-cddc5aa47952)

**Comment by cloudxxx8** [comment]
Store and Forward uses retry mechanism, so the service can't guarantee which event is sent out sucessfully first.
Also, the nature of message bus is asynchornous, the sending order won't be guaranteed.
https://stackoverflow.com/questions/30955110/is-message-order-preserved-in-mqtt-messages

The implementation is in https://github.com/edgexfoundry/app-functions-sdk-go

**Comment by amunozh** [comment]
Is there an easy way to guarantee ordering with Redis? Is it necessary to implement a custom storage?

**Reply by cloudxxx8** [reply]
You can implement the server side to sort the `origin` timestamp
